mds: add ephemeral pinning for subtrees

sidharthanup opened the pull request:
Fixes https://tracker.ceph.com/issues/41302 and https://tracker.ceph.com/issues/41541
Signed-off-by: Sidharth Anupkrishnan <sanupkri@redhat.com>
<!--
Thank you for opening a pull request!  Here are some tips on creating
a well formatted contribution.

Please give your pull request a title like "[component]: [short description]"

This is the format for commit messages:

"""
[component]: [short description]

[A longer multiline description]

Fixes: [ticket URL on tracker.ceph.com, create one if necessary]
Signed-off-by: [Your Name] <[your email]>
"""

The Signed-off-by line is important, and it is your certification that
your contributions satisfy the Developers Certificate or Origin.  For
more detail, see SubmittingPatches.rst.

The component is the short name of a major daemon or subsystem,
something like "mon", "osd", "mds", "rbd, "rgw", etc. For ceph-mgr modules,
give the component as "mgr/<module name>" rather than a path into pybind.

For more examples, simply use "git log" and look at some historical commits.

This was just a quick overview.  More information for contributors is available here:
https://raw.githubusercontent.com/ceph/ceph/master/SubmittingPatches.rst

-->
## Checklist
- [ ] References tracker ticket
- [ ] Updates documentation if necessary
- [ ] Includes tests for new functionality or reproducer for bug

---

<details>
<summary>Show available Jenkins commands</summary>

- `jenkins retest this please`
- `jenkins test crimson perf`
- `jenkins test signed`
- `jenkins test make check`
- `jenkins test make check arm64`
- `jenkins test submodules`
- `jenkins test dashboard`
- `jenkins test dashboard backend`
- `jenkins test docs`
- `jenkins render docs`

</details>

sidharthanup commented:
Added first few commits. Commits to come include:
1) How mds_map change is handled by the ranks.
2) Changes in balancer code.

batrick commented on src/mds/mdstypes.h:
```diff
@@ -636,7 +639,7 @@ struct inode_t {
 void inode_t<Allocator>::encode(bufferlist &bl, uint64_t features) const
 {
-  ENCODE_START(15, 6, bl);
+  ENCODE_START(17, 6, bl);
```
`16`

batrick commented on src/mds/mdstypes.h:
```diff
@@ -688,13 +691,16 @@ void inode_t<Allocator>::encode(bufferlist &bl, uint64_t features) const
 void inode_t<Allocator>::decode(bufferlist::const_iterator &p)
 {
-  DECODE_START_LEGACY_COMPAT_LEN(15, 6, 6, p);
+  DECODE_START_LEGACY_COMPAT_LEN(17, 6, 6, p);
```
`16` is fine.

batrick commented on src/mds/mdstypes.h:
```diff
@@ -783,6 +789,18 @@ void inode_t<Allocator>::decode(bufferlist::const_iterator &p)
+    decode(export_ephemeral_distributed_pin, p);
+  } else {
+    export_ephemeral_distributed_pin = MDS_RANK_NONE;
+  }
```
My issue with this change is that these values should not be in `inode_t`; these values should not be part of the struct written to directories in the metadata pool. I know we agreed they should be journaled because an MDS failover should not disrupt the pins but that only means it needs to be part of the journaled metadata. When the journal flushes to the metadata pool directory objects, these values are not persisted.

So what needs to happen is that this extra metadata needs to be part of a secondary structure which is included in `EMetaBlob::fullbit`. It'll be persisted in the journal events but not in the `inode_t`.

batrick replied:
I was partially confused when I wrote this comment. We **do** want to persist the vxattrs for random/distributed pin settings on a directory inode.

**Whether** the directory inode is ephemerally pinned will be stored in `EMetaBlob::fullbit`. Whether it's ephemerally pinned is indicated if it was randomly selected (the random ephemeral pin) or whether it is part of a distribution (the distributed ephemeral pin). Of course, **where** it's pinned is always determined by the consistent hash of its inode number, so how it was ephemerally pinned doesn't matter. So, we wouldn't expect to need to persist the rank it's pinned to in the journal. We can always recalculate it.

batrick commented on src/mds/MDSMap.h:
```diff
@@ -150,6 +150,21 @@ class MDSMap {
+  mds_rank_t put_ino_in_consistent_hash_ring(inodeno_t ino); //Assign the key i.e inode number to a rank in the consistent hash ring
+  vector<inodeno_t> get_inos_from_rank_in_consistent_hash_ring(mds_rank_t rank) const;
+  mds_rank_t get_successor_of_rank_in_consistent_hash_ring(mds_rank_t rank) const;
+  void remove_rank_node_from_consistent_hash_ring(mds_rank_t rank);
```
I have to pause my review at this point but I was surprised to find these methods here. I would expect most of this logic to exist in `MDCache`. Don't move anything yet as I'm not sure about this data structure either.

Let's talk about this during standup.

batrick commented on src/common/options.cc:
```diff
@@ -7954,6 +7954,16 @@ std::vector<Option> get_mds_options() {
     .set_default(true)
     .set_description("allow setting directory export pins to particular ranks"),
 
+    Option("mds_export_ephemeral_random", Option::TYPE_FLOAT, Option::LEVEL_ADVANCED)
```
`.set_flag(Option::FLAG_RUNTIME)`

batrick commented on src/common/options.cc:
```diff
@@ -7954,6 +7954,16 @@ std::vector<Option> get_mds_options() {
+    .set_long_description("probabilistically pin the loaded directory inode and the subtree beneath it to an MDS based on the consistent hash of the inode number. The higher this value the more likely the loaded subtrees get pinned"),
+   
+    Option("mds_export_ephemeral_distributed", Option::TYPE_BOOL, Option::LEVEL_ADVANCED)
+    .set_default(false)
```
`.set_flag(Option::FLAG_RUNTIME)`

batrick commented:
please rebase

batrick commented on src/mds/MDCache.cc:
```diff
@@ -829,7 +831,92 @@ MDSCacheObject *MDCache::get_object(const MDSCacheObjectInfo &info)
+
+void MDCache::add_rank_node_to_consistent_hash_ring(mds_rank_t rank)
+{
+  uint32_t hash = rjhash32(rank);
```
Why do ranks need to be hashed?

Logically the ranks are uniform "buckets" along the line (break the circle and stretch it out). Where each rank's bucket falls on that line is static based on the rank number and max_mds.

Since you're using `uint32_t` for the hash. If I had two ranks, rank 0 would get all inodes that hash between `[0, 2^31-1]` and rank 1 would get `[2^31, 2^32-1]`.

sidharthanup replied:
For the way I've implemented it:
The ranks in the consistent hash ring are arranged in the ascending order of the hash(rank). Technically this hash need not be saved as part of the struct (because it can be calculated anytime) but when an inode is hashed and the rank needs to be found out (for migration), the algorithm works by doing a binary search on the rank hashes to find the lowest rank hash that's greater than or equal to the inode hash (inspired by the consistent hashing defined in Akamai's paper). Since this involves constant comparison of the rank hashes, I thought of storing the hash in memory instead of recalculating it every time.

I think your suggestion is another way of achieving the same concept of partitioning the hash space by having a range assigned to each rank.

Note: I think I should do a modulo(max no of mdss that cephfs can support) to reduce the hash space more.

batrick replied:
I believe the calculation is literally as simple as:

````
mds_rank_t MDCache::hash_into_rank_bucket(uint32_t hash, mds_rank_t max_mds)
{
    return std::min(hash / (UINT32_MAX / (uint32_t)max_mds), max_mds);
}
````

> The ranks in the consistent hash ring are arranged in the ascending order of the hash(rank).

I haven't looked but I think the original paper is more complicated because it's not dealing with buckets that are consecutive integers without holes. We can safely use the above method instead.

sidharthanup replied:
Ahh! Got it now! The idea is the same as this paper https://arxiv.org/pdf/1406.2294.pdf I think.

sidharthanup replied:
@batrick When max_mds changes, won't all the keys in the hash table get redistributed in the hashing scheme you suggested? I think we should follow something relating to this: https://arxiv.org/pdf/1406.2294.pdf (Jump Consistent Hash).

batrick commented:
@sidharthanup status?

sidharthanup commented:
Removed the consistent_hash_ring implementation. As you suggested, since we are dealing with ranks that are consecutive integers and without holes, there is no need for a persist the data structures in memory. I have updated the hashing scheme similar to how it's done here : https://arxiv.org/pdf/1406.2294.pdf.

batrick commented on src/mds/CInode.cc:
```diff
@@ -5133,6 +5135,151 @@ void CInode::maybe_export_pin(bool update)
+{
+  //If both the configs aren't set then return
+  if (!g_conf().get_val<bool>("mds_export_ephemeral_random")
+      && !g_conf().get_val<bool>("mds_export_ephemeral_distributed"))
```
These configs need cached by in `MDCache`. Fetching these are relatively expensive. See `handle_conf_change`.

batrick commented on src/mds/CInode.cc:
```diff
@@ -5133,6 +5135,151 @@ void CInode::maybe_export_pin(bool update)
+
+  if (g_conf().get_val<bool>("mds_export_ephemeral_random")) {
+    if ((export_ephemeral_random_pin >=
+	  ceph::util::generate_random_number(0.0, 1.0) || update)
```
reverse these checks, `update` is cheaper to evaluate

batrick commented on src/mds/CInode.cc:
```diff
@@ -5133,6 +5135,151 @@ void CInode::maybe_export_pin(bool update)
+  if (g_conf().get_val<bool>("mds_export_ephemeral_random")) {
+    if ((export_ephemeral_random_pin >=
+	  ceph::util::generate_random_number(0.0, 1.0) || update)
+	&& export_ephemeral_distributed_pin == false) {
```
`!export...`?

batrick commented on src/mds/MDCache.cc:
```diff
@@ -829,6 +829,20 @@ MDSCacheObject *MDCache::get_object(const MDSCacheObjectInfo &info)
+    b = j;
+    hash = hash*2862933555777941757ULL + 1;
+    j = (b + 1) * (double(1LL << 31) / double((key >> 33) + 1));
+  }
```
Please explain this method more. It needs comments for the magic numbers especially.

markhpc replied:
@batrick  I believe this comes directly from Lamping & Veach.  See the implementation of JumpConsistentHash:

https://arxiv.org/pdf/1406.2294.pdf

batrick replied:
Sure, I suspected as much but it must the source must be cited and attribution given!

batrick commented on src/mds/CInode.cc:
```diff
@@ -5133,6 +5135,151 @@ void CInode::maybe_export_pin(bool update)
+        if (dir->is_subtree_root()) {
+          // set auxsubtree bit or export it
+          if (!dir->state_test(CDir::STATE_AUXSUBTREE) ||
+	      mdcache->hash_into_rank_bucket(ino(), mdcache->mds->mdsmap->get_max_mds()) != dir->get_dir_auth().first) {
```
Why is this not using the directory inode's ino number? i.e. why `ino()`?

sidharthanup replied:
ino() will return the inode number right?(https://github.com/ceph/ceph/blob/master/src/mds/CInode.h#L598)

batrick replied:
Ah, I see. I mistook this for an empty constructor.

batrick commented on src/mds/MDCache.cc:
```diff
@@ -13107,5 +13121,22 @@ void MDCache::handle_mdsmap(const MDSMap &mdsmap) {
-}
 
+  /* Handle consistent hash ring changes when new mds is active. In the case where an mds is stopped (max_mds is decreased), the subtree migrations are handled as the Inodes are loaded */
+  if (mdsmap.get_max_mds() > oldmap.get_max_mds()) {
```
Why not `!=`?

sidharthanup replied:
Because the `<` case is handled differently. When max_mds is reduced the inodes (that were a part of the stopped mds) that get loaded in other MDSs get hashed again and sent elsewhere.

batrick replied:
Thanks for adding a comment. From this:

> In the case where an mds is stopped (max_mds is decreased), the subtree migrations are handled as the Inodes are loaded

but I still don't understand. The inodes are already loaded. If we reduce max_mds, then inodes on the stopping MDS will need migrated to their new home. Current that's rank 0 IIRC. It is fine to maintain that (lazy but simple) behavior but keep in mind that once the rank stops, rank0 should be migrating the ephemerally pinned subtrees as required. I don't see how/where that's done in this code.

sidharthanup replied:
Ack. As it is now, the stopping MDS does not migrate the inodes that it handles (the inodes in its consistent hash list). The loaded inodes are dropped and once the MDS is stopped and there are only n-1 MDSs then when client requests come to the dir that was held by the now defunct nth MDS, rank 0 loads the metadata from RADOS, and
1) For ephemeral_random: Redoes the pinning probabilistically. May or may not get pinned again.
2) For ephemeral_distributed: Finds that the parent has ephemeral_distributed set and hence rehashes and redoes the pinning definitely.

I was worried about the delay in stopping time of the MDS, and other penalties if the migrations were done during the shutdown phase. Also in the current way, the loading is done when requests arrive and so is more dynamic(??). Do you think there is more penalties involved with reloading from RADOS, rather than migrations?

batrick replied:
> Ack. As it is now, the stopping MDS does not migrate the inodes that it handles (the inodes in its consistent hash list). The loaded inodes are dropped and once the MDS is stopped and there are only n-1 MDSs then when client requests come to the dir that was held by the now defunct nth MDS, rank 0 loads the metadata from RADOS, and

Not necessarily, IIRC. The stopping MDS needs to migrate any metadata that clients have capabilities for.

In any case, the code needs to consider shrinking max_mds. Consider `max_mds` shrinking from `3` to `2`:

````
|    0    |    1    |    2    |
|      0       |       1      |
````

Rank 1 clearly needs to migrate some of the metadata it has to rank 0.

>     1. For ephemeral_random: Redoes the pinning probabilistically. May or may not get pinned again.

Again, only true if the metadata is unpinned (no capabilities are issued, the metadata is unreplicated, etc.).

>     2. For ephemeral_distributed: Finds that the parent has ephemeral_distributed set and hence rehashes and redoes the pinning definitely.

Ditto.

> I was worried about the delay in stopping time of the MDS, and other penalties if the migrations were done during the shutdown phase.

Migrations are unavoidable. Note that in `MDCache::shutdown_pass`, the MDS tries to trip everything unpinned. Anything leftover must be exported or (eventually) expired if dirty.

> Also in the current way, the loading is done when requests arrive and so is more dynamic(??). Do you think there is more penalties involved with reloading from RADOS, rather than migrations?

Migrations would usually be cheaper in any case as the metadata will usually be warmer and loaded into the other MDS cache in batch.

sidharthanup replied:
Ack! I have changed it and added this logic in MDCache::shutdown_pass.

batrick replied:
No, it needs to be here too. All MDS ranks need to check their current ephemerally pinned inodes, not just the one shutting down.

batrick commented on src/mds/MDCache.h:
```diff
@@ -129,6 +129,8 @@ class MDCache {
   using clock = ceph::coarse_mono_clock;
   using time = ceph::coarse_mono_time;
 
+  vector<inodeno_t> consistent_hash_inodes;
```
This should be an `elist<CInode*>.

batrick replied:
You also don't remove inodes anywhere when they are dropped from the cache (migrated or dropped).

sidharthanup replied:
I didn't exactly understand this. Do you mean that when the inodes are trimmed, then it would be difficult to track the dropped inodes (consistent hashed ones) using just a vector of inode nos alone? I did fix the migration case (updating the vector) but the trimming part is interesting. Looking into it.

batrick replied:
You only need to do random insert, random erase, and iteration. That indicates a list is appropriate. (No lookup and no need for duplicate detection! So set is unnecessary.)

`elist` is an efficient way to create a list of `CInode`s without the overhead of `std::list`. It's used extensively in the MDS, just grep for it in `CInode.h`.

sidharthanup replied:
Ack. Still a bit unclear about this. Why do we need an elist of `CInodes` when all we need to track are the inode numbers?

batrick replied:
Because you need to do `get_inode` to fetch the `CInode` in the cache, which is extra overhead. Additionally, it makes it clear that anything in `consistent_hash_inodes` **must** be in the MDS cache. (The MDS should not care about pinned inodes on other ranks.)

sidharthanup replied:
Thanks! That is a reasonable justification. I will change the data structure to `elist`.

batrick commented on src/mds/Migrator.cc:
```diff
@@ -1114,8 +1114,12 @@ void Migrator::dispatch_export_dir(MDRequestRef& mdr, int count)
     // start the freeze, but hold it up with an auth_pin.
-    dir->freeze_tree();
-    ceph_assert(dir->is_freezing_tree());
+    //Ephemeral distributed pinning - Only the current (child) directory should get frozen and pinned
```
Pretty sure this isn't correct. Can you explain?

sidharthanup replied:
For ephemeral distributed, you shouldn't freeze the entire tree as freezing the entire tree implies the whole subtree gets migrated. The migrator exports whatever is frozen. So here we are freezing the directory only (not the entire subtree nested below)

batrick replied:
I'm not sure I follow. If ephemeral distributed is set on `/home` and we're going to migrate `/home/foo`, then `/home/foo` subtree should be frozen and migrated. It sounds like you're talking about `/home` should be frozen but that wouldn't be the case?

sidharthanup replied:
Great catch! Couldn't find this while debugging because there were no directories below in my test workload. i.e parent -> empty_dir so it got migrated anyways due to the `else` clause. Fixed it. Thanks!

batrick commented:
You also need to write tests for this. Please do this in `qa/tasks/cephfs/test_exports.py`. This should be easy to do with `vstart_runner.py`. If you have trouble with that, talk to Rishabh.

batrick commented:
Needs rebase.

ukernel commented on src/mds/mdstypes.h:
```diff
@@ -511,6 +511,8 @@ struct inode_t {
 
   mds_rank_t export_pin = MDS_RANK_NONE;
+
+  mds_rank_t export_ephemeral_random_pin = MDS_RANK_NONE;
```
why not reuse export_pin, define a special value for ephemeral_random_pin

ukernel commented on src/mds/CInode.cc:
```diff
@@ -5133,6 +5134,76 @@ void CInode::maybe_export_pin(bool update)
+    return;
+
+  if (g_conf()->export_ephemeral_random) {
+    if (g_conf()->mds_export_ephemeral_random >= ceph::util::generate_random_number(0.0, 1.0)) {
```
why do we need to persistent the target rank.

sidharthanup commented:
@batrick @ukernel Rebased and distributed the code among fresh commits. @ukernel Please check again as the code had changed drastically from the old commits.

ukernel commented on src/mds/CInode.cc:
```diff
@@ -5176,6 +5182,111 @@ void CInode::maybe_export_pin(bool update)
+        queue = true;
+        }
+        if (queue) {
+	  if (mdcache->hash_into_rank_bucket(ino(), mdcache->mds->mdsmap->get_max_mds()) == mdcache->mds->get_nodeid()) {
```
what does this mean?

sidharthanup replied:
Each rank has its own data structure (consistent_hash_inodes) to keep a track of the inodes that it handles as part of the consistent hash ring. If the inode no hashes to the same rank as the current mds (i.e no migration is required), then insert it into the data structure. The data structure is else populated during migration on the importer's side during import.

ukernel replied:
what does 'mdcache->hash_into_rank_bucket(ino(), ....' means

sidharthanup replied:
This method hashes the inode no into a rank.. almost according to a consistent hashing strategy specified in this paper: https://arxiv.org/pdf/1406.2294.pdf (Jump Consistent Hash)

ukernel commented on src/mds/CInode.cc:
```diff
@@ -5176,6 +5182,111 @@ void CInode::maybe_export_pin(bool update)
+  if (mdcache->get_export_ephemeral_random_config()) {
+    if ((update || export_ephemeral_random_pin >=
+	  ceph::util::generate_random_number(0.0, 1.0))
+	&& is_export_ephemeral_distributed_pinned == false) {
```
If we set random pin on a directory, all descendant directories under the directory will  be randomly distributed to all mds. This may create lots of subtrees.  lot of subtrees add lots of overhead on mds journal. I don't know what 's the use case of ephemeral_random_pin

sidharthanup replied:
The whole subtree below it is supposed to be migrated to a random rank. It is hierarchical like export_pin. So the implementation is almost the same.

ukernel replied:
export_ephemeral_random_pin is from  get_export_ephemeral_random_pin(), which return true if any ancestor dir has export_ephemeral_random_pin enabled .  This means inodes under an export_ephemeral_random_pin enabled directory will be randomly distributed

batrick replied:
In practice, `export_ephemeral_random_pin` will be something like 1% or maybe significantly less. It would be rare for a child directory to be ephemerlaly pinned.

ukernel replied:
Iwhat benefits we can get from export_ephemeral_random_pin?

batrick replied:
Because the pinning is hierarchical, it's somewhat likely that a randomly ephemerally pinned cold directory will still be a parent of a hot directory.

What real gains we see from this is still unknown. I think it warrants research and experimentation after it's done.

ukernel replied:
unless a hot directory is in very deep path, the hot directory is still unlikely to be distributed. I still don't think export_ephemeral_random_pin is useful.

batrick replied:
That may be, we'll see!

ukernel commented on src/mds/CDir.cc:
```diff
@@ -3154,6 +3154,13 @@ void CDir::_freeze_dir()
+    ceph_assert(auth.first >= 0);
+    ceph_assert(auth.second == CDIR_AUTH_UNKNOWN);
+    auth.second = auth.first;
+  }
```
I don't think this change and most changes in Migrator.cc are not needed

sidharthanup replied:
Ack dropped all of these. I had thought that only the child directory and not the entire child subtree needed to be exported. Fixed it.

sidharthanup commented:
@batrick and @ukernel Please check again. Fixed the issues.

markhpc commented:
@sidharthanup I gave this a spin while testing the io500 benchmark on an nvme backed cluster.  The "mdtest_easy_write" portion of the io500 creates a directory for every mpi process of mdtest being run (on this cluster there are 10 nodes and 20 processes per node).  This seemed to be an ideal case for this test as I'm using 80 MDS to try and achieve a high degree of parallelism.  With this PR in place performance was actually worse than master.  ceph fs status shows very uneven distribution of inos and requests across MDS instances.  Is there some tuning that I need to perform beyond the defaults?

```
+------+--------+-----+---------------+-------+-------+
| Rank | State  | MDS |    Activity   |  dns  |  inos |
+------+--------+-----+---------------+-------+-------+
|  0   | active |  a  | Reqs: 3984 /s |  655k |  655k |
|  1   | active |  b  | Reqs:    0 /s |  134k |  134k |
|  2   | active |  c  | Reqs:    0 /s |  227k |  227k |
|  3   | active |  d  | Reqs:    0 /s |  202k |  202k |
|  4   | active |  e  | Reqs:    0 /s |  141k |  141k |
|  5   | active |  f  | Reqs: 12.5k/s |  370k |  370k |
|  6   | active |  g  | Reqs:    0 /s |  188k |  188k |
|  7   | active |  h  | Reqs: 9337 /s |  140k |  140k |
|  8   | active |  i  | Reqs: 7475 /s | 82.7k | 82.7k |
|  9   | active |  j  | Reqs: 2733 /s | 26.2k | 26.2k |
|  10  | active |  k  | Reqs:    0 /s | 37.1k | 37.1k |
|  11  | active |  l  | Reqs:    0 /s | 89.9k | 89.9k |
|  12  | active |  m  | Reqs:    0 /s | 31.0k | 31.0k |
|  13  | active |  n  | Reqs:    0 /s | 10.0k | 10.0k |
|  14  | active |  o  | Reqs:    0 /s | 9178  | 9181  |
|  15  | active |  p  | Reqs:    0 /s | 18.7k | 18.7k |
|  16  | active |  q  | Reqs:    0 /s | 54.7k | 54.7k |
|  17  | active |  r  | Reqs:    0 /s | 22.8k | 22.8k |
|  18  | active |  s  | Reqs:    0 /s | 21.2k | 21.2k |
|  19  | active |  t  | Reqs:    0 /s | 34.7k | 34.7k |
|  20  | active |  u  | Reqs:    0 /s | 4478  | 4481  |
|  21  | active |  v  | Reqs:    0 /s |  133k |  133k |
|  22  | active |  w  | Reqs:    0 /s | 23.9k | 23.9k |
|  23  | active |  x  | Reqs:    0 /s | 4753  | 4756  |
|  24  | active |  y  | Reqs:    0 /s | 13.9k | 13.9k |
|  25  | active |  z  | Reqs:    0 /s | 13.5k | 13.5k |
|  26  | active |  aa | Reqs:    0 /s | 11.0k | 11.0k |
|  27  | active |  ab | Reqs:    0 /s | 4612  | 4615  |
|  28  | active |  ac | Reqs: 8467 /s |  290k |  290k |
|  29  | active |  ad | Reqs:    0 /s |  149k |  149k |
|  30  | active |  ae | Reqs:    0 /s |  291k |  291k |
|  31  | active |  af | Reqs:    0 /s |  153k |  153k |
|  32  | active |  ag | Reqs: 11.8k/s |  221k |  221k |
|  33  | active |  ah | Reqs:    0 /s | 9022  | 9025  |
|  34  | active |  ai | Reqs:    0 /s | 8315  | 8318  |
|  35  | active |  aj | Reqs:    0 /s | 7354  | 7357  |
|  36  | active |  ak | Reqs:    0 /s | 9977  | 9980  |
|  37  | active |  al | Reqs:    0 /s | 11.4k | 11.4k |
|  38  | active |  am | Reqs:    0 /s | 18.8k | 18.8k |
|  39  | active |  an | Reqs:    0 /s | 9226  | 9229  |
|  40  | active |  ao | Reqs:    0 /s | 4575  | 4578  |
|  41  | active |  ap | Reqs:    0 /s | 9624  | 9627  |
|  42  | active |  aq | Reqs:    0 /s | 14.0k | 14.0k |
|  43  | active |  ar | Reqs:    0 /s |   16  |   19  |
|  44  | active |  as | Reqs:    0 /s | 47.9k | 47.9k |
|  45  | active |  at | Reqs:    0 /s | 19.1k | 19.1k |
|  46  | active |  au | Reqs:    0 /s | 19.1k | 19.1k |
|  47  | active |  av | Reqs:  212 /s | 5376  | 5379  |
|  48  | active |  aw | Reqs:  968 /s | 27.4k | 27.4k |
|  49  | active |  ax | Reqs:  898 /s | 27.4k | 27.4k |
|  50  | active |  ay | Reqs:    0 /s | 52.2k | 52.2k |
|  51  | active |  az | Reqs:    0 /s |  110k |  110k |
|  52  | active |  ba | Reqs:  946 /s | 31.8k | 31.8k |
|  53  | active |  bb | Reqs:    0 /s |   16  |   19  |
|  54  | active |  bc | Reqs:    0 /s |   16  |   19  |
|  55  | active |  bd | Reqs:    0 /s | 4559  | 4562  |
|  56  | active |  be | Reqs:    0 /s | 9658  | 9661  |
|  57  | active |  bf | Reqs:    0 /s | 23.7k | 23.7k |
|  58  | active |  bg | Reqs:    0 /s | 9088  | 9091  |
|  59  | active |  bh | Reqs:    0 /s | 28.2k | 28.2k |
|  60  | active |  bi | Reqs:    0 /s |   16  |   19  |
|  61  | active |  bj | Reqs:    0 /s |   16  |   19  |
|  62  | active |  bk | Reqs:    0 /s |   16  |   19  |
|  63  | active |  bl | Reqs:    0 /s | 4693  | 4696  |
|  64  | active |  bm | Reqs:    0 /s |   16  |   19  |
|  65  | active |  bn | Reqs:    0 /s | 33.8k | 33.8k |
|  66  | active |  bo | Reqs:    0 /s | 1852  | 1855  |
|  67  | active |  bp | Reqs:    0 /s | 14.3k | 14.3k |
|  68  | active |  bq | Reqs:    0 /s | 3686  | 3689  |
|  69  | active |  br | Reqs:    0 /s | 3686  | 3689  |
|  70  | active |  bs | Reqs:    0 /s |   16  |   19  |
|  71  | active |  bt | Reqs:    0 /s | 11.0k | 11.0k |
|  72  | active |  bu | Reqs:    0 /s | 5520  | 5523  |
|  73  | active |  bv | Reqs:    0 /s | 1852  | 1855  |
|  74  | active |  bw | Reqs:    0 /s | 18.6k | 18.6k |
|  75  | active |  bx | Reqs:    0 /s | 24.6k | 24.6k |
|  76  | active |  by | Reqs:    0 /s | 22.9k | 22.9k |
|  77  | active |  bz | Reqs:    0 /s | 23.0k | 23.0k |
|  78  | active |  ca | Reqs:    0 /s | 7354  | 7357  |
|  79  | active |  cb | Reqs:    0 /s | 5520  | 5523  |
+------+--------+-----+---------------+-------+-------+
```

sidharthanup commented:
@markhpc Like subtree pinning, this PR introduces xattrs that the cluster admin has to set on directories to modify the metadata/subtree distribution across MDSs. So running a workload which doesn't explicitly set these xattrs as the directories are getting created, shouldn't cause anything new and as such shouldn't deviate from the master, which in your run I suppose it is due the metadata balancer's variable exporting behavior (@batrick ?).

markhpc commented:
@sidharthanup I reran tests using random pinning (it would be really nice if getting the xattr after it's been set worked so you could verify the change actually takes place!).  I did in fact see much better ino distribution but no real performance improvement.  I'll try reading through the PR today and see if I can run some more tests to figure out what's going on.

```
+------+--------+-----+---------------+-------+-------+
| Rank | State  | MDS |    Activity   |  dns  |  inos |
+------+--------+-----+---------------+-------+-------+
|  0   | active |  a  | Reqs:   37 /s |  275k |  275k |
|  1   | active |  b  | Reqs:    0 /s |  306k |  306k |
|  2   | active |  c  | Reqs:    0 /s |  332k |  332k |
|  3   | active |  d  | Reqs:    0 /s |  353k |  353k |
|  4   | active |  e  | Reqs:    0 /s |  236k |  236k |
|  5   | active |  f  | Reqs:    0 /s |  295k |  295k |
|  6   | active |  g  | Reqs:    0 /s |  241k |  241k |
|  7   | active |  h  | Reqs:    0 /s |  172k |  172k |
|  8   | active |  i  | Reqs:    0 /s |  191k |  191k |
|  9   | active |  j  | Reqs:    0 /s |  356k |  356k |
|  10  | active |  k  | Reqs:    0 /s |  241k |  241k |
|  11  | active |  l  | Reqs: 1097 /s |  366k |  366k |
|  12  | active |  m  | Reqs:    0 /s |  234k |  234k |
|  13  | active |  n  | Reqs:    0 /s |  149k |  149k |
|  14  | active |  o  | Reqs:    0 /s |  200k |  200k |
|  15  | active |  p  | Reqs:    0 /s |  208k |  208k |
|  16  | active |  q  | Reqs:    0 /s |  197k |  197k |
|  17  | active |  r  | Reqs:   75 /s |  279k |  279k |
|  18  | active |  s  | Reqs:    0 /s |  309k |  309k |
|  19  | active |  t  | Reqs:    0 /s |  308k |  308k |
|  20  | active |  u  | Reqs: 1519 /s |  294k |  294k |
|  21  | active |  v  | Reqs: 1905 /s |  252k |  252k |
|  22  | active |  w  | Reqs:  628 /s |  191k |  191k |
|  23  | active |  x  | Reqs:  121 /s |  314k |  314k |
|  24  | active |  y  | Reqs:    0 /s |  158k |  158k |
|  25  | active |  z  | Reqs:    0 /s |  260k |  260k |
|  26  | active |  aa | Reqs: 1173 /s |  339k |  339k |
|  27  | active |  ab | Reqs:    0 /s |  170k |  170k |
|  28  | active |  ac | Reqs:  877 /s |  163k |  163k |
|  29  | active |  ad | Reqs:  312 /s |  267k |  267k |
|  30  | active |  ae | Reqs:    0 /s |  148k |  148k |
|  31  | active |  af | Reqs:  145 /s |  198k |  198k |
|  32  | active |  ag | Reqs:    0 /s |  187k |  187k |
|  33  | active |  ah | Reqs:  618 /s |  191k |  191k |
|  34  | active |  ai | Reqs:    0 /s |  246k |  246k |
|  35  | active |  aj | Reqs:    0 /s |  197k |  197k |
|  36  | active |  ak | Reqs:  936 /s |  178k |  178k |
|  37  | active |  al | Reqs:    0 /s |  166k |  166k |
|  38  | active |  am | Reqs:  297 /s |  275k |  275k |
|  39  | active |  an | Reqs:    0 /s |  219k |  219k |
|  40  | active |  ao | Reqs: 1029 /s |  191k |  191k |
|  41  | active |  ap | Reqs:    0 /s |  135k |  135k |
|  42  | active |  aq | Reqs:  600 /s |  197k |  197k |
|  43  | active |  ar | Reqs:    0 /s |  155k |  155k |
|  44  | active |  as | Reqs:    0 /s |  192k |  192k |
|  45  | active |  at | Reqs:  226 /s |  204k |  204k |
|  46  | active |  au | Reqs:    0 /s |  174k |  174k |
|  47  | active |  av | Reqs:    0 /s |  187k |  187k |
|  48  | active |  aw | Reqs:    0 /s |  294k |  294k |
|  49  | active |  ax | Reqs:    0 /s |  126k |  126k |
|  50  | active |  ay | Reqs:    0 /s |  283k |  283k |
|  51  | active |  az | Reqs:    0 /s |  194k |  194k |
|  52  | active |  ba | Reqs:    0 /s |  160k |  160k |
|  53  | active |  bb | Reqs:    0 /s |  137k |  137k |
|  54  | active |  bc | Reqs: 3258 /s |  333k |  333k |
|  55  | active |  bd | Reqs:  153 /s |  120k |  120k |
|  56  | active |  be | Reqs:    0 /s |  129k |  129k |
|  57  | active |  bf | Reqs: 1594 /s |  189k |  189k |
|  58  | active |  bg | Reqs:    0 /s |  192k |  192k |
|  59  | active |  bh | Reqs:  937 /s |  168k |  168k |
|  60  | active |  bi | Reqs:    0 /s |  128k |  128k |
|  61  | active |  bj | Reqs:    0 /s |  137k |  137k |
|  62  | active |  bk | Reqs:    0 /s |  247k |  247k |
|  63  | active |  bl | Reqs:    0 /s |  170k |  170k |
|  64  | active |  bm | Reqs:    0 /s |  174k |  174k |
|  65  | active |  bn | Reqs:    0 /s |  151k |  151k |
|  66  | active |  bo | Reqs:    0 /s |  165k |  165k |
|  67  | active |  bp | Reqs:    0 /s | 78.1k | 78.1k |
|  68  | active |  bq | Reqs:    0 /s |  228k |  228k |
|  69  | active |  br | Reqs:    0 /s |  176k |  176k |
|  70  | active |  bs | Reqs:    0 /s |  491k |  491k |
|  71  | active |  bt | Reqs:    0 /s |  137k |  137k |
|  72  | active |  bu | Reqs:    0 /s |  132k |  132k |
|  73  | active |  bv | Reqs:    0 /s |  102k |  102k |
|  74  | active |  bw | Reqs:    0 /s |  137k |  137k |
|  75  | active |  bx | Reqs:    0 /s |  170k |  170k |
|  76  | active |  by | Reqs:    0 /s | 83.1k | 83.1k |
|  77  | active |  bz | Reqs:    0 /s |  148k |  148k |
|  78  | active |  ca | Reqs:    0 /s |  135k |  135k |
|  79  | active |  cb | Reqs:    0 /s |  143k |  143k |
+------+--------+-----+---------------+-------+-------+
```

batrick commented on src/mds/MDBalancer.cc:
```diff
@@ -99,7 +99,13 @@ void MDBalancer::handle_export_pins(void)
+    // Making sure the ephemeral pin does not override export pin
+    if (in->get_export_pin(false) != MDS_RANK_NONE)
+      export_pin = in->get_export_pin(false);
+    else if (in->is_export_ephemeral_distributed_migrating || in->is_export_ephemeral_random_migrating)
```
You're missing `{}` here which causes a bug: (non-ephemeral) export pins are always overridden.

Coding lesson: always use `{}` for control statements. The LOC saved is never worth it and you may hit this kind of coding error whenever you add another line.

markhpc commented on src/mds/CInode.cc:
```diff
@@ -5176,6 +5183,117 @@ void CInode::maybe_export_pin(bool update)
+            queue = true;
+        } else {
+        // create aux subtree or export it
+        queue = true;
```
Needs more indentation

batrick commented:
please rebase

sidharthanup commented:
@markhpc Thanks for the review! Apologize for the indentation nits. The tabs/spaces mixup messed it up. Please do check again.

batrick commented on src/mds/MDCache.cc:
```diff
@@ -830,6 +840,23 @@ MDSCacheObject *MDCache::get_object(const MDSCacheObjectInfo &info)
+// consistent hash ring
+
+/*
+ * hashing implementation based on Lamping and Veach's Jump Consistent Hash: https://arxiv.org/pdf/1406.2294.pdf
```
> based on

Did you make changes, or is it a verbatim copy? What changes did you make?

batrick replied:
Why is this algorithm useful for this work? Explain how it is used to implement consistent hashing for ephemeral pins.

batrick commented on qa/tasks/cephfs/test_exports.py:
```diff
@@ -175,3 +175,47 @@ def test_session_race(self):
+        self.assertGreaterEqual(rank1_distributed_subtree_ratio, 0.2)
+
+        rank2_distributed_subtree_ratio = len(self.get_distributed_auth_subtrees(status, 2)/len(self.get_auth_subtrees(status, 2))
+        self.assertGreaterEqual(rank2_distributed_subtree_ratio, 0.2)
```
Test should stop here.

Add another test for growing max_mds. Add a third test for shrinking max_mds.

batrick commented on qa/tasks/cephfs/test_exports.py:
```diff
@@ -175,3 +175,47 @@ def test_session_race(self):
+        total_ephemerally_pinned_auth_subtree_count = rank0_ephemerally_pinned_auth_subtree_count + rank1_ephemerally_pinned_auth_subtree_count + rank2_ephemerally_pinned_auth_subtree_count
+        self.fs.set_max_mds(2)
+        self.fs.wait_for_daemons()
+        time.sleep(15)
```
We try to avoid using `time.sleep` in testing. Use `self.wait_until_true`

batrick commented on src/mds/CInode.cc:
```diff
@@ -5212,6 +5330,46 @@ mds_rank_t CInode::get_export_pin(bool inherit) const
+      break;
+    // Check if the inode is projected/dirty
+    if (in->is_projected() && in->get_projected_inode()->export_ephemeral_random_pin >= 0)
+      return in->get_projected_inode()->export_ephemeral_random_pin;
```
Don't look at the projected value. Use the same logic as export pins. See fix here: 93cbee3dd9c1446a027e7673645a1856ea68637a

batrick commented on src/mds/CInode.cc:
```diff
@@ -2031,6 +2034,8 @@ void CInode::decode_lock_ipolicy(bufferlist::const_iterator& p)
     decode(inode.export_pin, p);
     maybe_export_pin(old_pin != inode.export_pin);
+    decode(inode.export_ephemeral_distributed_pin, p);
+    decode(inode.export_ephemeral_random_pin, p);
```
If this changes, we need to call `maybe_ephemeral_pin`.

batrick replied:
Ditto in `pop_and_dirty_projected_inode`

batrick commented on src/mds/Server.cc:
```diff
@@ -5625,6 +5625,50 @@ void Server::handle_set_vxattr(MDRequestRef& mdr, CInode *cur)
     auto &pi = cur->project_inode();
     cur->set_export_pin(rank);
     pip = &pi.inode;
+  }  else if (name.find("ceph.dir.random.pin") == 0) {
```
rename `ceph.dir.pin.random`

batrick commented on src/mds/Server.cc:
```diff
@@ -5625,6 +5625,50 @@ void Server::handle_set_vxattr(MDRequestRef& mdr, CInode *cur)
+    auto &pi = cur->project_inode();
+    cur->set_export_ephemeral_random_pin(val);
+    pip = &pi.inode;
+  } else if (name.find("ceph.dir.distributed.pin") == 0) {
```
`ceph.dir.pin.distributed`

batrick commented on src/mds/MDCache.cc:
```diff
@@ -13256,5 +13300,31 @@ void MDCache::handle_mdsmap(const MDSMap &mdsmap) {
+  /* Handle consistent hash ring changes when new mds is active. In the case where an mds is stopped (max_mds is decreased), the subtree migrations are handled as the Inodes are loaded */
+  if (mdsmap.get_max_mds() > oldmap.get_max_mds()) {
+    if (export_ephemeral_random_config ||
+        export_ephemeral_distributed_config) {
```
No need to gate this here.

batrick commented on src/mds/MDCache.cc:
```diff
@@ -217,6 +221,10 @@ void MDCache::handle_conf_change(const std::set<std::string>& changed, const MDS
+  if (changed.count("mds_export_ephemeral_distributed"))
+    export_ephemeral_distributed_config = g_conf().get_val<bool>("mds_export_ephemeral_distributed");
+  if (changed.count("mds_export_ephemeral_random"))
+    export_ephemeral_random_config = g_conf().get_val<bool>("mds_export_ephemeral_random");
```
If one of these are disabled, the MDS should go through its ephemerally pinned subtrees and unpin them.

batrick commented:
(Trying to keep these reviews small as GitHub has lost my reviews in the past...)

batrick commented on src/mds/CInode.cc:
```diff
@@ -5178,6 +5189,126 @@ void CInode::maybe_export_pin(bool update)
+
+  if (export_ephemeral_distributed_config) {
+
+    CDentry *pdn = get_projected_parent_dn();
```
`get_parent_dn()`?

batrick commented on src/mds/CInode.cc:
```diff
@@ -454,6 +455,8 @@ void CInode::pop_and_dirty_projected_inode(LogSegment *ls)
   if (new_export_pin)
     maybe_export_pin(true);
 
+  if (front.inode.version == 1)
```
This needs an comment explanation.

batrick commented on src/mds/MDCache.cc:
```diff
@@ -7788,6 +7817,23 @@ bool MDCache::shutdown_pass()
     migrator->clear_export_queue();
+
+    if (export_ephemeral_random_config ||
+        export_ephemeral_distributed_config) {
```
I think we can drop this `if` check.

batrick commented on src/mds/MDCache.cc:
```diff
@@ -6723,9 +6753,8 @@ std::pair<bool, uint64_t> MDCache::trim(uint64_t count)
-	migrator->export_empty_import(dir);
-        ++trimmed;
+        migrator->export_empty_import(dir);
+	++trimmed;
```
please undo this change

batrick commented on src/mds/MDCache.cc:
```diff
@@ -13256,5 +13302,30 @@ void MDCache::handle_mdsmap(const MDSMap &mdsmap) {
+
+  /* Handle consistent hash ring during cluster resizes */
+  if (mdsmap.get_max_mds() != oldmap.get_max_mds()) {
+        dout(10) << "Max mds is increased. Noted from MDCache side" << dendl;
```
this comment needs updated (max_mds may be higher or lower)

batrick replied:
This and many debug messages need updated. Instead of just noting what happens, say what the code is going to do about it. For example, this would be better:

`dout(10) << "Checking ephemerally pinned directories for re-export due to max_mds change." << dendl;`

batrick commented on src/mds/MDCache.cc:
```diff
@@ -13256,5 +13302,30 @@ void MDCache::handle_mdsmap(const MDSMap &mdsmap) {
+            }
+            else {
+              dout(10) << "adding inode to export queue" << dendl;
+              (*it)->maybe_export_ephemeraldistributed_pin(true);
```
typo

batrick commented on src/mds/MDCache.cc:
```diff
@@ -13256,5 +13302,30 @@ void MDCache::handle_mdsmap(const MDSMap &mdsmap) {
+        dout(10) << "Max mds is increased. Noted from MDCache side" << dendl;
+        elist<CInode*>::iterator it = ephemeral_pins.begin(member_offset(CInode, ephemeral_pin_inode));
+        while (!it.end()) {
+          // Migrate if the inodes hash elsewhere
```
`auto in = *it;`

simplifies all the code below

batrick commented on src/mds/MDCache.cc:
```diff
@@ -13256,5 +13302,30 @@ void MDCache::handle_mdsmap(const MDSMap &mdsmap) {
+        elist<CInode*>::iterator it = ephemeral_pins.begin(member_offset(CInode, ephemeral_pin_inode));
+        while (!it.end()) {
+          // Migrate if the inodes hash elsewhere
+          if (hash_into_rank_bucket((*it)->ino(), mdsmap.get_max_mds()) != mds->get_nodeid()) {
```
The duplication of logic here with `maybe_export_ephemeral_*` should be avoidable. Let's make a method:

`CInode::queue_ephemeral_export()`

That takes an inode that **is** ephemerally pinned already and tries to export it to the right place. If it's already in the right place, it removes it from the queue.

batrick commented:
@sidharthanup I am making amendments to your commits. Please push any new changes as new commits and I will squash them.

batrick commented:
Also note: your last update didn't compile. I'm trying to help fix that.

batrick commented:
@sidharthanup I see you did a rebase of your own. I don't want to break anything you did so I put my branch here:

https://github.com/batrick/ceph/tree/pr30592

It is getting closer. Distributed pinning mostly works now.

sidharthanup commented:
@batrick Ack. I haven't changed anything other than the tests though.

gregsfortytwo commented:
@sidharthanup, if "Fix test issues" is just straightforward fixes to the tests in "qa: add tests for ephemeral pinning", can you please squash it down?

If not, it'll need to get a more proper title eg "qa: resolve various issues in ephemeral pinning tasks" and a broader explanation of what was wrong and why the changes make it right.

gregsfortytwo commented:
Ping @sidharthanup.

Also I merged this into master and there's a simple conflict in test_exports. Will run some QA tests since it looks like that hasn't been done in the lab in quite a while.

sidharthanup commented:
@gregsfortytwo That test failure was what I was talking about in the standup. After debugging, I figured out the reason was this commit: https://github.com/ceph/ceph/pull/30592/commits/17c0de81d7daf8756adeac8fea4f051bcb1f1370 . The check would prevent empty directories that were meant to be pinned in other mds. Ideally export pinned empty directories should be exported as the importer does not send it back: https://github.com/ceph/ceph/blob/master/src/mds/Migrator.cc#L793. I have added a commit to fix it.

sidharthanup commented:
http://pulpito.ceph.com/sidharthanup-2020-03-14_12:45:04-multimds-wip-sidharth-testing-2020-03-14-1606-distro-default-smithi/

sidharthanup commented:
Today's multimds run: http://pulpito.ceph.com/sidharthanup-2020-03-16_09:25:42-multimds-wip-sidharth-testing-2020-03-14-1606-distro-default-smithi/ . Seems ok

gregsfortytwo commented:
@sidharthanup Awesome. We flag an error when a standby takes over because it's often a sign of performance issues we need to deal with. Did you look and see why the failover occurred? Since it's happening with an export test we need to make sure we didn't do something in this PR to make it slower.

sidharthanup commented:
@gregsfortytwo Its due to this test case: https://github.com/ceph/ceph/pull/30592/files#diff-d5f17ebd745250b57be2b89d4ba48efbR459 . Here we purposefully kill an mds and let a standby take over it's place to make sure that the ephemerally pinned subtrees are retained on journal replay. I have scheduled a run skipping this particular test case: http://pulpito.ceph.com/sidharthanup-2020-03-17_13:33:41-multimds-wip-sidharth-testing-2020-03-14-1606-distro-default-smithi/ .

sidharthanup commented:
@gregsfortytwo Is there any way to not consider that run as an error? (Since the standby taking over is the main focus of that test case)

gregsfortytwo commented:
> @gregsfortytwo Is there any way to not consider that run as an error? (Since the standby taking over is the main focus of that test case)

Yes! You need to add that log entry to the whitelist in the test yaml. There are many examples in the qa suites to make sure you get the syntax right. :)

sidharthanup commented:
@gregsfortytwo Thanks! Looking into that now.

sidharthanup commented:
http://pulpito.ceph.com/sidharthanup-2020-03-17_21:01:41-multimds-wip-sidharth-testing-2020-03-18-0113-distro-default-smithi/

gregsfortytwo commented:
@sidharthanup, the tests are not actually running:

2020-03-17T21:22:14.135 INFO:tasks.cephfs_test_runner:test_ephemeral_pin_dist_off (tasks.cephfs.test_exports.TestEphemeralPins) ... skipped 'Only have 3 MDSs, require 4'

sidharthanup commented:
@greg I don't understand why that's happening. Isn't teuthology supposed to bring up the N no of MDS according to MDS_REQUIRED? My earlier runs didn't have this problem. 

Scheduled another run here: http://pulpito.ceph.com/sidharthanup-2020-03-18_07:49:56-multimds-wip-sidharth-testing-2020-03-18-0113-distro-default-smithi/

sidharthanup commented:
Yup got it! So when multimds jobs are run there are two possible configurations - one is using 3-mds.yaml which runs it with 3 mdss and 9-mds.yaml which runs it with 9 mdss. The tests get skipped when run with the first config. In http://pulpito.ceph.com/sidharthanup-2020-03-17_21:01:41-multimds-wip-sidharth-testing-2020-03-18-0113-distro-default-smithi/ , three of those jobs are with the first configuration (where the tests are skipped) and two of those jobs (job id : 4864100 and 4864102) the tests are run and pass

gregsfortytwo commented:
@sidharthanup Unfortunately I do't think we're set up to handle this the way you want to with some nice flag. If it's only test_exports that needs 4 nodes, we'll need to make sure that task is only run against the 9-MDS config by changing the files or their arrangement somehow.

For instance we could duplicate the existing multimds/basic suite infrastructure into multimds/basic/all-sizes and multimds/basic/large-count, and have all-sizes run against both 3 and 9 clusters but the large-count only run against 9, and then separate the tasks into those two suites as appropriate. I don't think we want to leave a mismatch in the tree though since it'll waste machine time and give us false confidence in test successes.

sidharthanup commented:
@greg Ack. In that case I have slightly modified the test cases. The only test cases where you actually need 4 MDS were test_ephemeral_pin_grow_mds (where max_mds was changed from 3 to 4) and test_ephemeral_standby (where max_mds was 3 and killing an MDS triggered the 4th standby MDS to take rank 2) and both of those test cases can be changed to use a maximum of 3 MDSs and not 4. Have amended the commit.

sidharthanup commented:
http://pulpito.ceph.com/sidharthanup-2020-03-19_10:41:16-multimds-wip-sidharth-testing-2020-03-18-0113-distro-default-smithi/ . All tests pass.

sidharthanup commented:
@gregsfortytwo Please take a look.

sidharthanup commented:
Fixed base branch to master and rebased to resolve conflict. @gregsfortytwo Please check.

batrick commented on src/mds/MDCache.cc:
```diff
@@ -10814,6 +10883,7 @@ void MDCache::decode_replica_inode(CInode *&in, bufferlist::const_iterator& p, C
   inodeno_t ino;
   snapid_t last;
   __u32 nonce;
+  unsigned mask;
```
bump decode version above

batrick commented on src/mds/MDCache.cc:
```diff
@@ -10714,6 +10780,9 @@ void MDCache::encode_replica_inode(CInode *in, mds_rank_t to, bufferlist& bl,
   in->_encode_base(bl, features);
   in->_encode_locks_state_for_replica(bl, mds->get_state() < MDSMap::STATE_ACTIVE);
+  if (in->state_test(CInode::STATE_RANDEPHEMERALPIN)) {
+    encode(CInode::STATE_RANDEPHEMERALPIN, bl);
```
bump encode version above:

`ENCODE_START(2, 1, bl);`

batrick commented on src/mds/MDCache.h:
```diff
@@ -215,6 +215,18 @@ class MDCache {
+
+  void ephemeral_pin(elist<CInode*>::item& in) {
+    ephemeral_pins.push_back(&in);
+  }
```
I'm confused how this change reappeared. It's causing build failures.

sidharthanup replied:
I think I messed it up when I rebased after that merge conflict. I'm cleaning everything up now

batrick commented on src/mds/MDCache.cc:
```diff
@@ -10910,7 +10914,16 @@ void MDCache::decode_replica_inode(CInode *&in, bufferlist::const_iterator& p, C
   }
-  DECODE_FINISH(p); 
+
+  if (!p.end()) {
```
`if (struct_v >= 2) {`

batrick commented:
Also: please break the last commit into separate logical commits with expalanations.

ukernel commented:
This PR defines separate sets of functions (in CInode.cc) for dist pin and random pin. Now there are three sets of similar functions. I'd like to unify the codes that handle export_pin,  maybe_ephemeral_dist_pin and maybe_ephemeral_random_pin.

batrick commented:
> This PR defines separate sets of functions (in CInode.cc) for dist pin and random pin. Now there are three sets of similar functions. I'd like to unify the codes that handle export_pin, maybe_ephemeral_dist_pin and maybe_ephemeral_random_pin.

Let's save it for a future refactor. We need to finalize this PR.

ukernel commented:
Once a  non-empty directory get exported by ephemeral_random_pin, it will stay in mds cache forever. We need a way to handle this.

ukernel commented:
I still doubt benefit of ephemeral random pin. I'd like to see benchmark result.

sidharthanup commented:
> Once a non-empty directory get exported by ephemeral_random_pin, it will stay in mds cache forever. We need a way to handle this.
@ukernel  Why do you think it will stay in the cache forever and not get trimmed?

sidharthanup commented:
@batrick Cases I don't think we've handled properly yet are:
1)  switching from ephemeral dist to ephemeral rand on a dir/dir_tree. Need to handle state changes (CInode::STATE_DISTEPHEMERALPIN and CInode::STATE_RANDEPHEMERALPIN). And also changing both the inode xattrs as well.
2) switching from export_pin to ephemeral_dist/ephemeral_random and vice versa - the inode xattr for only one of them (the one being set) is changed and persisted while the other one remains intact.

Do you think these are necessary?

batrick commented:
> @batrick Cases I don't think we've handled properly yet are:
> 
>     1. switching from ephemeral dist to ephemeral rand on a dir/dir_tree. Need to handle state changes (CInode::STATE_DISTEPHEMERALPIN and CInode::STATE_RANDEPHEMERALPIN). 

What state change? I don't follow.

> And also changing(and persisting) both the inode xattrs(rather than just one) as well.

Both xattrs are persisted via `inode_t`. What do you mean?

>     2. switching from export_pin to ephemeral_dist/ephemeral_random and vice versa - the inode xattr for only one of them (the one being set) is changed and persisted while the other one remains intact.

I think it makes some sense to have a default like:

````
setfattr -n ceph.dir.pin -v 0 /cephfs
setfattr -n ceph.dir.pin.random -v .01 /cephfs
````

In which case, the least surprising result is that all directories are pinned to rank 0 by default but there is a one percent chance a new dir or one loaded from the meta pool is pinned elsewhere.

sidharthanup commented:
> > @batrick Cases I don't think we've handled properly yet are:
> > ```
> > 1. switching from ephemeral dist to ephemeral rand on a dir/dir_tree. Need to handle state changes (CInode::STATE_DISTEPHEMERALPIN and CInode::STATE_RANDEPHEMERALPIN). 
> > ```
> 
> What state change? I don't follow.

If we are going to set CInode::State_DISTEPHEMERALPIN on an inode, check if CInode::STATE_RANDEPHEMERALPIN is already set and if so then unset it and then only set CInode::State_DISTEPHEMERALPIN. The same for the vice-versa case. Shouldn't this be checked.

> > And also changing(and persisting) both the inode xattrs(rather than just one) as well.
> 
> Both xattrs are persisted via `inode_t`. What do you mean?
> 
> > ```
> > 2. switching from export_pin to ephemeral_dist/ephemeral_random and vice versa - the inode xattr for only one of them (the one being set) is changed and persisted while the other one remains intact.

I was referring to the case where we set the ephemeral_dist xattr on a dir and then later set ephemeral_rand on that same directory. In this case the ephemeral_dist xattr is left unchanged in inode_t when ephemeral_rand is set to a value in inode_t. Shouldn't ephemeral_dist xattr be unset (i.e set to 0) as well, in this case (and same for the vice versa case) : 
```
setfattr -n ceph.dir.pin.random -v .01 /cephfs
setfattr -n ceph.dir.pin.dist -v 1 /cephfs
```

Of course, these problems wouldn't arise if the user unsets the xattr and then proceeds to set the other. i.e : 
```
setfattr -n ceph.dir.pin.random -v .01 /cephfs
setfattr -n ceph.dir.pin.random -v 0 /cephfs
setfattr -n ceph.dir.pin.dist -v 1 /cephfs
```
Is it expected of the user to unset the xattr before setting the other one?

> I think it makes some sense to have a default like:
> 
> ```
> setfattr -n ceph.dir.pin -v 0 /cephfs
> setfattr -n ceph.dir.pin.random -v .01 /cephfs
> ```
> 
> In which case, the least surprising result is that all directories are pinned to rank 0 by default but there is a one percent chance a new dir or one loaded from the meta pool is pinned elsewhere.

Agreed!

batrick commented:
> > > @batrick Cases I don't think we've handled properly yet are:
> > > ```
> > > 1. switching from ephemeral dist to ephemeral rand on a dir/dir_tree. Need to handle state changes (CInode::STATE_DISTEPHEMERALPIN and CInode::STATE_RANDEPHEMERALPIN). 
> > > ```
> > 
> > 
> > What state change? I don't follow.
> 
> If we are going to set CInode::State_DISTEPHEMERALPIN on an inode, check if CInode::STATE_RANDEPHEMERALPIN is already set and if so then unset it and then only set CInode::State_DISTEPHEMERALPIN. The same for the vice-versa case. Shouldn't this be checked.

What scenario do you envision where it doesn't make sense for both states to be set? I don't see an issue if this is the case. It does not change the logic of the ephemeral pin.

> > > And also changing(and persisting) both the inode xattrs(rather than just one) as well.
> > 
> > 
> > Both xattrs are persisted via `inode_t`. What do you mean?
> > > ```
> > > 2. switching from export_pin to ephemeral_dist/ephemeral_random and vice versa - the inode xattr for only one of them (the one being set) is changed and persisted while the other one remains intact.
> > > ```
> 
> I was referring to the case where we set the ephemeral_dist xattr on a dir and then later set ephemeral_rand on that same directory. In this case the ephemeral_dist xattr is left unchanged in inode_t when ephemeral_rand is set to a value in inode_t. Shouldn't ephemeral_dist xattr be unset (i.e set to 0) as well, in this case (and same for the vice versa case) :
> 
> ```
> setfattr -n ceph.dir.pin.random -v .01 /cephfs
> setfattr -n ceph.dir.pin.dist -v 1 /cephfs
> ```

I don't see them as disjoint policies. You can have both.

> Of course, these problems wouldn't arise if the user unsets the xattr and then proceeds to set the other. i.e :
> 
> ```
> setfattr -n ceph.dir.pin.random -v .01 /cephfs
> setfattr -n ceph.dir.pin.random -v 0 /cephfs
> setfattr -n ceph.dir.pin.dist -v 1 /cephfs
> ```
> 
> Is it expected of the user to unset the xattr before setting the other one?

No.

> > I think it makes some sense to have a default like:
> > ```
> > setfattr -n ceph.dir.pin -v 0 /cephfs
> > setfattr -n ceph.dir.pin.random -v .01 /cephfs
> > ```
> > 
> > 
> > In which case, the least surprising result is that all directories are pinned to rank 0 by default but there is a one percent chance a new dir or one loaded from the meta pool is pinned elsewhere.
> 
> Agreed!

sidharthanup commented:
@batrick Yea your'e right. On second thought, they don't have to be mutually exclusive.

ukernel commented:
> > Once a non-empty directory get exported by ephemeral_random_pin, it will stay in mds cache forever. We need a way to handle this.
> 
> @ukernel Why do you think it will stay in the cache forever and not get trimmed?

because balancer does not export dirfrag back to dir inode's auth mds. MDS can't trim subtree roots.

sidharthanup commented:
> > > Once a non-empty directory get exported by ephemeral_random_pin, it will stay in mds cache forever. We need a way to handle this.
> > 
> > 
> > @ukernel Why do you think it will stay in the cache forever and not get trimmed?
> 
> because balancer does not export dirfrag back to dir inode's auth mds. MDS can't trim subtree roots.

Ahh! Thanks for the explanation!
But this is default balancer/trimming behaviour for subtree partioning right? Every dir that gets exported will be a subtree root on the importer mds, and it's dir inode's auth is the exporter mds. And these auth subtree roots can't be trimmed. Why should this be handled just for for ephemeral random pinning?

ukernel commented:
On Fri, May 29, 2020 at 3:42 AM Sidharth Anupkrishnan <
notifications@github.com> wrote:

> Once a non-empty directory get exported by ephemeral_random_pin, it will
> stay in mds cache forever. We need a way to handle this.
>
> @ukernel <https://github.com/ukernel> Why do you think it will stay in
> the cache forever and not get trimmed?
>
> because balancer does not export dirfrag back to dir inode's auth mds. MDS
> can't trim subtree roots.
>
> Ahh! Thanks for the explanation!
> But this is default balancer/trimming behaviour for subtree partioning
> right? Every dir that gets exported will be a subtree root on the importer
> mds, and it's dir inode's auth is the exporter mds. And these auth subtree
> roots can't be trimmed. Why should this be handled just for for ephemeral
> random pinning?
>
> 1. If there is no pin, balancer will export idle subtree root back to dir
inode's auth mds.
2. ephemeral random pinning will create more sutrees than balancer creates.




> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/ceph/ceph/pull/30592#issuecomment-635556122>, or
> unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AAVFTL6QFWT336HNK26HW73RT25BXANCNFSM4I2Z6PQA>
> .
> [image: image.gif]
>
>

batrick commented:
@sidharthanup I'm messing with the commits in this PR so don't push anything new.

What's the rationale for:

`mds: Fix migration of empty directories when the directory is export`

Did

`mds: skip export of empty directories`

break something?

sidharthanup commented:
> @sidharthanup I'm messing with the commits in this PR so don't push anything new.

Ack
> What's the rationale for:
> 
> `mds: Fix migration of empty directories when the directory is export`
> 
> Did
> 
> `mds: skip export of empty directories`
> 
> break something?

Yes. `mds: skip export of empty directories` caused one of the asserts in TestExports (not TestEphemeralPins) in test_exports.py to fail. This was because the check in the commit prevented empty directories that had the export_pin applied to it, to be migrated, and so the desired subtree state wasn't achieved. Note: The importer doesn't send back empty export pinned directories by default: https://github.com/ceph/ceph/blob/master/src/mds/Migrator.cc#L793. My commit just added leeway for empty export_pinned directories to continue their default behavior.

batrick commented:
> > What's the rationale for:
> > `mds: Fix migration of empty directories when the directory is export`
> > Did
> > `mds: skip export of empty directories`
> > break something?
> 
> Yes. `mds: skip export of empty directories` caused one of the asserts in TestExports (not TestEphemeralPins) in test_exports.py to fail. This was because the check in the commit prevented empty directories that had the export_pin applied to it, to be migrated, and so the desired subtree state wasn't achieved. Note: The importer doesn't send back empty export pinned directories by default: https://github.com/ceph/ceph/blob/master/src/mds/Migrator.cc#L793. My commit just added leeway for empty export_pinned directories to continue their default behavior.

Thanks for the explanation. I've decided to change this behavior and fixed the test. Otherwise it makes dropping the pinned subtree from cache difficult or impossible.

batrick commented:
@sidharthanup I've pushed my changes. Please have a look.

sidharthanup commented on src/mds/CInode.h:
```diff
@@ -325,14 +325,22 @@ class CInode : public MDSCacheObject, public InodeStoreBase, public Counter<CIno
+  /* These are for "permanent" state markers that are passed around between
+   * MDS. Nothing protects/updates it like a typicl MDS lock.
+   */
+  static const int MASK_STATE_REPLICATED = STATE_RANDEPHEMERALPIN;
```
Can you give more info about what purpose `MASK_STATE_REPLICATED` serves for ephemeral_random pinning?

sidharthanup commented on src/mds/MDCache.cc:
```diff
@@ -10865,6 +10945,16 @@ void MDCache::decode_replica_inode(CInode *&in, bufferlist::const_iterator& p, C
+  if (struct_v >= 2) {
+    __u32 state;
+    decode(state, p);
+    state &= CInode::MASK_STATE_REPLICATED;
```
Why can't we just bitwise & with `CInode::STATE_RANDEPHEMERALPIN` . I didn't get why `CInode::MASK_STATE_REPLICATED` is defined separately (but using the value of `CInode::STATE_RANDEPHEMERALPIN`)?

batrick replied:
> Why can't we just bitwise & with `CInode::STATE_RANDEPHEMERALPIN`

I'm following the same logic as for the exported/import case, e.g. `CInode::decode_import`. We can't just `state_set` though for random pins because it needs added to the random pin set as well. That work is all done in the `set_ephemeral_rand`.

sidharthanup replied:
Ack. I wasn't questioning the logic though. Just was slightly confused why define and use CInode::MASK_STATE_REPLICATED when we already have CInode::STATE_RANDEPHEMERALPIN. But yea, it doesn't sound good to use CInode::STATE_RANDEPHEMERALPIN as a placeholder for replicated inode.

sidharthanup commented on src/mds/CInode.h:
```diff
@@ -325,14 +325,22 @@ class CInode : public MDSCacheObject, public InodeStoreBase, public Counter<CIno
+     STATE_DISTEPHEMERALPIN|STATE_RANDEPHEMERALPIN);
+
+  /* These are for "permanent" state markers that are passed around between
+   * MDS. Nothing protects/updates it like a typicl MDS lock.
```
s/typicl/typical/

sidharthanup commented on src/mds/MDCache.cc:
```diff
@@ -10735,6 +10809,12 @@ void MDCache::encode_replica_inode(CInode *in, mds_rank_t to, bufferlist& bl,
   in->_encode_base(bl, features);
   in->_encode_locks_state_for_replica(bl, mds->get_state() < MDSMap::STATE_ACTIVE);
+
+  if (struct_v >= 2) {
+    __u32 state = in->state;
+    encode(state, bl);
+  }
```
Can it not be done conditionally? i.e encode state only if ephemeral rand is set on that inode?

batrick replied:
No, because the importer wouldn't be able to tell if it's present or not. Consider when we add more to this bufferlist in the future. We cant have arbitrary holes in the encoding.

sidharthanup replied:
Ack. This reasoning is a good justification.

sidharthanup commented:
jenkins render docs

ceph-jenkins commented:
Doc render available at http://docs.ceph.com/ceph-prs/30592/

sidharthanup commented on doc/cephfs/multimds.rst:
```diff
@@ -135,3 +135,102 @@ directory's export pin. For example:
+
+Both random and distributed ephemeral pin policies are off by default in
+Octopus. The features may be enabled via the
+``mds_export_ephemeral_random_max`` and ``mds_export_ephemeral_distributed``
```
s/mds_export_ephemeral_random_max/mds_export_ephemeral_random

batrick replied:
fixed, thanks!

sidharthanup replied:
@batrick The change hasn't been reflected here yet.

batrick replied:
Thanks. I thought I had fixedc it but the change go5t lost somehow.

sidharthanup commented on qa/tasks/cephfs/test_exports.py:
```diff
@@ -12,121 +12,126 @@ class TestExports(CephFSTestCase):
-        self.mount_a.setfattr("1/2", "ceph.dir.pin", "0")
-        self._wait_subtrees([('/1', 1), ('/1/2', 0)], status=status, rank=1)
-        self._wait_subtrees([('/1', 1), ('/1/2', 0)], status=status)
+        # Check if rank1 changed (standby tookover?)
```
Shouldn't MDS_REQUIRED (on the top) be 3 instead of 2 because of this change?

sidharthanup replied:
Disregard this comment. This wouldn't matter because teuthology schedules 3 MDSs

sidharthanup commented:
@batrick Last commit's commit message should be break up export_pin tests right?

sidharthanup commented:
Everything else LGTM!

batrick commented:
> @batrick Last commit's commit message should be break up export_pin tests right?

Yes, thanks!

batrick commented on src/pybind/mgr/volumes/fs/operations/pin_util.py:
```diff
@@ -0,0 +1,36 @@
+from distutils.util import strtobool
+
+_pin_value = {
+    "export": lambda x: int(strtobool(x)),
```
Whoops, this should only be an int. Good catch!

batrick commented on src/pybind/mgr/volumes/fs/operations/pin_util.py:
```diff
@@ -0,0 +1,36 @@
+
+_pin_value = {
+    "export": lambda x: int(strtobool(x)),
+    "distributed": lambda x: int(strtobool(x)),
```
distributed actually is any bool so we do want this.

batrick commented on src/pybind/mgr/volumes/fs/operations/pin_util.py:
```diff
@@ -0,0 +1,36 @@
+
+    try:
+        pin_setting = _pin_value[pin_type](pin_setting)
+    except Exception as e:
```
good point

ajarr commented:
Fixes tracker https://tracker.ceph.com/issues/41541 ?

batrick commented:
http://pulpito.ceph.com/?branch=wip-pdonnell-testing-20200623.022209

Failures exist in master. Reduced data availability warnings look to be a new OSD issue. One known failure:

https://tracker.ceph.com/issues/46023

fsstress failures are known issues as well.

batrick commented:
Going to resolve the tox error then merge.

batrick commented:
jenkins test docs

batrick commented:
jenkins test docs
